README.md
# A collection of useful scripts around Elite Dangerous #

## Requirements ##

* argcomplete
* argparse
* requests
* Tkinter

You probably want to install your distribution/OS package for Tkinter instead of
using pip. Then do the good old `pip install --user -r requirements.txt`.

You’ll also need to install pyEDSM’s dependencies:
`pip install --user -r pyEDSM/requirements.txt`

## Pre-Compiled Version (Windows) ##

If you check the
[releases](https://github.com/alterNERDtive/elite-scripts/releases) you’ll find
a .zip file with pre-compiled scripts that don’t need Python installed. Mostly
a) for saving myself some hassle and b) because that’s easier to distribute with
my
[VoiceAttack profiles](https://github.com/alterNERDtive/VoiceAttack-profiles).

## Scripts ##

### edsm-getnearest.py ###

```
usage: edsm-getnearest.py [-h] --system SYSTEM [--short] [--gui | --text]
                          CMDR [CMDR ...]

Locate your CMDRs using EDSM and find their distance to a given system.

positional arguments:
  CMDR             a list of CMDR names (must have their location public on
                   EDSM!)

options:
  -h, --help       show this help message and exit
  --system SYSTEM  the target system (must be in EDDN!)
  --short          short output (only makes sense with `--text`)
  --gui            explicitly run the GUI
  --text           explicitly give text output
```

### edts.py ###

```
usage: edts.py [-h] {coords} ...

Script for interfacing with Alot’s hosted EDTS API.

options:
  -h, --help  show this help message and exit

subcommands:
  {coords}    sub-command help
    coords    Searches for the approximate coordinates of a given procedurally
              generated system name.
```

```
usage: edts.py coords [-h] [--maxuncertainty [MAXUNCERTAINTY]] system

positional arguments:
  system                the system name to get coordinates for

options:
  -h, --help            show this help message and exit
  --maxuncertainty [MAXUNCERTAINTY]
                        maximum accepted uncertainty, if any
```

### explorationtools.py ###

```
usage: explorationtools.py [-h]
                           {bodycount,distancebetween,findcommander,findsystem,systemlist}
                           ...

A collection of tools useful for exploration.

options:
  -h, --help            show this help message and exit

subcommands:
  {bodycount,distancebetween,findcommander,findsystem,systemlist}
                        sub-command help
    bodycount           Returns the number of bodies in a system. Will exit
                        with code 1 on server error and code 2 if the system
                        could not be found in EDSM.
    distancebetween     Calculates the distance between two systems. Will exit
                        with code 1 on server error and code 2 if (one of) the
                        systems could not be found on EDSM.
    findcommander       Attempts to find a CMDR’s last known position. Will
                        exit with code 1 on server error and code 2 if the
                        CMDR could not be found on EDSM. Will also give you
                        the time of last activity if you search for their
                        system.
    findsystem          Attempts to find a partially matching system that
                        should then hopefully be in the vicinity of the given
                        system
    systemlist          Pulls all system names starting with the given string
                        from EDSM
```

```
usage: explorationtools.py bodycount [-h] system

positional arguments:
  system      system to query

options:
  -h, --help  show this help message and exit
```

```
usage: explorationtools.py distancebetween [-h] [--roundto [ROUNDTO]]
                                           system system

positional arguments:
  system               the systems to measure

options:
  -h, --help           show this help message and exit
  --roundto [ROUNDTO]  the number of digits to round to (default: 2)
```

```
usage: explorationtools.py findcommander [-h] [--system | --coords | --url]
                                         name [apikey]

positional arguments:
  name        the commander in question
  apikey      the commander’s EDSM API key. Can be empty for public profiles.

options:
  -h, --help  show this help message and exit
  --system    output the commander’s last known system (default)
  --coords    output the commander’s last known position in {x,y,z}
              coordinates
  --url       output the commander’s profile URL
```

```
usage: explorationtools.py findsystem [-h] system

positional arguments:
  system      the system in question

options:
  -h, --help  show this help message and exit
```

```
usage: explorationtools.py systemlist [-h] partialsystem

positional arguments:
  partialsystem  the partial system name to query against

options:
  -h, --help     show this help message and exit
```

### spansh.py ###

```
usage: spansh.py [-h] [--raw] {nearestsystem,oldstations,systemexists} ...

Script for interfacing with Spansh’s API.

options:
  -h, --help            show this help message and exit
  --raw                 output raw json (mostly for debugging)

subcommands:
  {nearestsystem,oldstations,systemexists}
                        sub-command help
    nearestsystem       Searches for the nearest system in the database to
                        given coordinates.
    oldstations         Searches for stations with old data (>1 year without
                        an update.
    systemexists        Checks if a given system exists in the search
                        database.
```

```
usage: spansh.py nearestsystem [-h] [--short | --parsable]
                               coordinate coordinate coordinate

positional arguments:
  coordinate  the coordinates to search for (order: x, y, z)

options:
  -h, --help  show this help message and exit
  --short     short output format (system name only)
  --parsable  parsable output format (<name>|<x>,<y>,<z>|<distance>)
```

```
usage: spansh.py oldstations [-h] [--system [SYSTEM]] [--count [COUNT]]
                             [--minage [MINAGE]] [--short] [--nofeet]

options:
  -h, --help         show this help message and exit
  --system [SYSTEM]  a single system to query. If not present, get the oldest
                     stations overall.
  --count [COUNT]    how many stations to output. Defaults to 50.
  --minage [MINAGE]  minimum age of data (in days) to be considered
                     “outdated”. Defaults to 365 (= 1 year).
  --short            short output format (system/station names only)
  --nofeet           omit Odyssey settlements
```

```
usage: spansh.py systemexists [-h] system

positional arguments:
  system      the system to search for

options:
  -h, --help  show this help message and exit
```

## Need Help / Want to Contribute? ##

Just [file an issue](https://github.com/alterNERDtive/elite-scripts/issues/new)
here or [hop into Discord](https://discord.gg/XHNX7jN) if that is your thing.


CHANGELOG.md
# devel

## Added

* `spansh.py`: `--raw` option. Prints the raw JSON returned from a query.
* `spansh.py`: `--nofeet` option for `outdatedstations`. Excludes Odyssey 
  settlements from the result set.

## Changed

* `spansh.py`: Now uses <https://spansh.uk> instead of <https://spansh.co.uk>.

-----

# 0.7 (2021-01-28)

## Changed

* `spansh.py`: Fleet carriers are now explicitly excluded from the outdated 
  stations list.

## Fixed

* `explorationtools.py`: Searching for a CMDR with public profile, but hidden 
  activity will no longer error out.
* `explorationtools.py`: Searching for a CMDR with public profile, but hidden 
  flight log will no longer error out.

-----

# 0.6 (2020-07-18)

## Added

* `spansh.py` old station snearch now takes a `--minage` (in days) argument for 
  what you would consider “outdated” data.

## Changed

* `spansh.py` old station search now outputs an age in days for the last update 
  instead of a time stamp.

-----

# 0.5 (2020-07-01)

Changed Changelog format. Should be even clearer now at a glance!

See [KeepAChangelog](https://keepachangelog.com/en/1.0.0/).

## Added

* `edts.py`: Script for accessing Alot’s hosted EDTS API. Right now the only 
  implemented thing is getting approximate coordinates for a given procedually 
  generated system name.

-----

# 0.4.1 (2020-06-18)

* removed accidentally left over debug print code from `spansh.py`

-----

# 0.4 (2020-03-09)

Kind of a big one. Obviously because in addition to EDSM, I can now do some 
stuff with Spansh’s database. At this point I’m not sure if I want to keep it in 
the kind of hacky quick&dirty state it’s in right now or do another abstraction 
layer in addition to EDSM’s. Probably will be too lazy.

In addition to that the compiled version of the scripts for windows will now be 
a single executable file per script instead of having all the libraries stuff in 
subfolders. It’s cleaner that way, IMO.

## explorationtools.py

* `findcommander` now gives a last seen date for the position

known bugs:

* I need to fix the way the commander’s last seen is pulled. Right now searching 
  for a commander will error out if their profile is public, but their flight 
  logs (and thus the last updated time stamp) is not.

## spansh.py

* added `spansh.py` for accessing data from https://spansh.co.uk
* see the README for what it can do right now


edts.py
#!/usr/bin/env python3
# PYTHON_ARGCOMPLETE_OK
import argcomplete, argparse
import json as JSON
import requests
import urllib
import sys

from pyEDSM.edsm.exception import ServerError, NotFoundError

# ===========================================================================

class ProcGenNameError(Exception):
  def __str__(self):
    return "Given system is not a valid proc gen name."

def getCoords(system):
  url = APIURLS["coords"] + urllib.parse.quote(system)
  response = requests.get(url)
  if response.status_code != 200:
    if response.status_code == 400:
      raise ProcGenNameError()
    else:
      raise ServerError(url, {})

  json = response.json()['result']

  ret = json['position']
  ret['uncertainty'] = json['uncertainty']

  return ret

# ===========================================================================

parser = argparse.ArgumentParser(description="Script for interfacing with "
    + "Alot’s hosted EDTS API.")
subparsers = parser.add_subparsers(title="subcommands", help="sub-command help",
    dest="subcommand", required=True)

parser_coords = subparsers.add_parser("coords",
    help="Searches for the approximate coordinates of a given procedurally "
      + "generated system name.")
parser_coords.add_argument("system",
    help="the system name to get coordinates for")
parser_coords.add_argument("--maxuncertainty", nargs="?", type=int,
    help="maximum accepted uncertainty, if any")

argcomplete.autocomplete(parser)
args = parser.parse_args()

# ===========================================================================

APIURLS = {
    "coords": "http://edts.thargoid.space/api/v1/system_position/"
  }

try:
  if args.subcommand == "coords":
    coords = getCoords(args.system)
    if args.maxuncertainty:
      if args.maxuncertainty < coords['uncertainty']:
        raise NotFoundError()
    out = "{},{},{}|{}".format(int(coords['x']), int(coords['y']),
        int(coords['z']), int(coords['uncertainty']))
except ServerError as e:
  print(e)
  sys.exit(1)
except ProcGenNameError as e:
  print(e)
  sys.exit(4)
except NotFoundError as e:
  print("Maximum uncertainty exceeded: " + str(int(coords['uncertainty']))
      + " > " + str(args.maxuncertainty))
  sys.exit(3)
else:
  print(out)
  sys.exit(0)


generate_docs.sh
#!/usr/bin/env bash
cat > README.md << EOF
# A collection of useful scripts around Elite Dangerous #

## Requirements ##

* argcomplete
* argparse
* requests
* Tkinter

You probably want to install your distribution/OS package for Tkinter instead of
using pip. Then do the good old \`pip install --user -r requirements.txt\`.

You’ll also need to install pyEDSM’s dependencies:
\`pip install --user -r pyEDSM/requirements.txt\`

## Pre-Compiled Version (Windows) ##

If you check the
[releases](https://github.com/alterNERDtive/elite-scripts/releases) you’ll find
a .zip file with pre-compiled scripts that don’t need Python installed. Mostly
a) for saving myself some hassle and b) because that’s easier to distribute with
my
[VoiceAttack profiles](https://github.com/alterNERDtive/VoiceAttack-profiles).

## Scripts ##

### edsm-getnearest.py ###

\`\`\`
EOF
./edsm-getnearest.py -h >> README.md
cat >> README.md << EOF
\`\`\`

### edts.py ###

\`\`\`
EOF
./edts.py -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./edts.py coords -h >> README.md
cat >> README.md << EOF
\`\`\`

### explorationtools.py ###

\`\`\`
EOF
./explorationtools.py -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./explorationtools.py bodycount -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./explorationtools.py distancebetween -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./explorationtools.py findcommander -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./explorationtools.py findsystem -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./explorationtools.py systemlist -h >> README.md
cat >> README.md << EOF
\`\`\`

### spansh.py ###

\`\`\`
EOF
./spansh.py -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./spansh.py nearestsystem -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./spansh.py oldstations -h >> README.md
cat >> README.md << EOF
\`\`\`

\`\`\`
EOF
./spansh.py systemexists -h >> README.md
cat >> README.md << EOF
\`\`\`

## Need Help / Want to Contribute? ##

Just [file an issue](https://github.com/alterNERDtive/elite-scripts/issues/new)
here or [hop into Discord](https://discord.gg/XHNX7jN) if that is your thing.
EOF


edsm-getnearest.py
#!/usr/bin/env python3
import argparse
import math
import requests
import sys
import tkinter as tk

from pyEDSM.edsm.exception import CommanderNotFoundError, ServerError, SystemNotFoundError
from pyEDSM.edsm.models import Commander, System

# =================================================================================

def getDistances (system, cmdrs, roundTo=2):
  distances = {}
  for cmdr in cmdrs:
    distances[cmdr] = system.distanceTo(cmdr, roundTo=roundTo)
  return distances

# =================================================================================

def outputGui():
  def runsearch(event=None):
    for child in frame.winfo_children():
      child.grid_remove()
      child.destroy()
    try:
      system = System(systemField.get())
      distances = getDistances(system, cmdrs)
      nearestCmdr = min(distances,key=distances.get)
      lbl = tk.Label(
          frame, text='nearest CMDR: {} ({} ly from {})'.format(nearestCmdr.name,
            distances[nearestCmdr], system.name))
      lbl.grid(row=0, columnspan=2)
      row = 1
      for cmdr in distances:
        row += 1
        lbl = tk.Label(frame, text='{}:'.format(cmdr.name))
        lbl.grid(row=row, column=0)
        lbl = tk.Label(frame, text='{} ly'.format(distances[cmdr]))
        lbl.grid(row=row, column=1)
    except (ServerError, SystemNotFoundError) as e:
      lbl = tk.Label(frame, text=e)
      lbl.grid(row=0, columnspan=2)
    except EdsmApiException as e:
      lbl = tk.Label(frame, text=e)
      lbl.grid(row=0, columnspan=2)
  window = tk.Tk()
  window.title('EDSM nearest CMDR')
  lbl = tk.Label(window, text='system:')
  lbl.grid(row=0, column=0)
  systemField = tk.Entry(window, width=50)
  systemField.grid(row=0, column=1)
  systemField.insert(tk.END, system.name)
  systemField.focus()
  frame = tk.Frame(window)
  frame.grid(row=1, columnspan=3)
  btn = tk.Button(window, text='get distances', command=runsearch)
  btn.grid(row=0, column=2)
  window.bind('<Return>', runsearch)
  runsearch()
  window.mainloop()

# =================================================================================

def outputText():
  if shortOutput:
    roundTo=0
  else:
    roundTo=2
  try:
    distances = getDistances(system, cmdrs, roundTo=roundTo)
  except CommanderNotFoundError as e:
    print(e)
    sys.exit(1)
  except ServerError as e:
    print(e)
    sys.exit(1)
  except SystemNotFoundError as e:
    print(e)
    sys.exit(2)
  nearestCmdr = min(distances,key=distances.get)
  if shortOutput:
    print('nearest commander: {} ({} ly).'.format(nearestCmdr.name,
      int(distances[nearestCmdr])))
  else:
    print('nearest CMDR: {} ({} ly from {}).'.format(nearestCmdr.name,
      distances[nearestCmdr], system.name))
    print()
    for cmdr in distances:
      print('{}: {} ly'.format(cmdr.name, distances[cmdr]))
    sys.exit(0)

# =================================================================================

parser = argparse.ArgumentParser(description='Locate your CMDRs using EDSM and '
    + 'find their distance to a given system.')
parser.add_argument('cmdrs', metavar='CMDR', nargs='+', help='a list of CMDR names '
    + '(must have their location public on EDSM!)')
parser.add_argument('--system', nargs=1, help='the target system (must be in '
    + 'EDDN!)', required=True)
parser.add_argument('--short', action='store_true', help='short output (only '
    + 'makes sense with `--text`)')
group = parser.add_mutually_exclusive_group()
group.add_argument('--gui', action='store_true', help='explicitly run the GUI')
group.add_argument('--text', action='store_true', help='explicitly give text output')

args = parser.parse_args()

system = System(args.system[0].strip().replace(' ', '').replace('', ''))
cmdrs = []
for name in args.cmdrs:
  cmdrs += [Commander(name)]
shortOutput = args.short

# =================================================================================

if args.text:
  outputText()
elif args.gui:
  outputGui()
else:
  try:
    outputGui()
  except tk.TclError:
    outputText()


explorationtools.py
#!/usr/bin/env python3
# PYTHON_ARGCOMPLETE_OK
import argcomplete, argparse
import math
import sys
from datetime import datetime

from pyEDSM.edsm.exception import ServerError, NotFoundError
from pyEDSM.edsm.models import System, Commander

# ===========================================================================

def getBodyCount(system):
  return System(system).bodyCount

def distanceBetween(system1, system2, roundTo=2):
  systems = System.getSystems(system1, system2)
  distance = systems[0].distanceTo(systems[1], roundTo)
  if roundTo == 0:
    distance = int(distance)
  return distance

def getCommanderPosition(name, apikey):
  coords = Commander(name, apikey).currentPosition
  ret = "hidden"
  if coords:
    ret = ""
    for k in coords:
      ret += "{}: {}, ".format(k, coords[k])
    ret = ret[:-2]
  return ret

def getCommanderProfileUrl(name, apikey):
  return Commander(name, apikey).profileUrl

def getCommanderSystem(name, apikey):
  cmdr = Commander(name, apikey)
  if cmdr.lastActivity is None:
    return "{}".format(cmdr.currentSystem)
  else:
    return "{} (last seen {})".format(cmdr.currentSystem,
        when(cmdr.lastActivity))
def when(date):
  diff = datetime.now() - date
  ret = ""
  if diff.days > 0:
    ret += "{} days ".format(diff.days)
  if diff.seconds > 0:
    hours = int(diff.seconds / 3600)
    if hours > 0:
      ret += "{} hours ".format(hours)
    minutes = int(diff.seconds % 3600 / 60)
    if minutes > 0:
      ret += "{} minutes ".format(minutes)
    if diff.days == 0 and hours == 0 and minutes == 0:
      # ONLY seconds!
      ret = "{} seconds ".format(diff.seconds)
  ret += "ago"
  return ret

def getSystemList(name):
  systems = System.getSystems(name)
  ret = ""
  for system in systems:
    ret += "{}\n".format(system.name)
  return ret[:-1]

def getSystemNear(name):
  # Probably want to abort at _some_ point. I’m defining two full words left as
  # the condition for that now.
  if name.count(' ') < 2:
    ret = "Aborting search at {}, require more than 2 words to limit the "
    ret += "result set."
    return ret.format(name)

  try:
    systems = System.getSystems(name)
  except NotFoundError:
    return getSystemNear(name[:-1])
  else:
    ret = ""
    for system in systems:
      ret += "{} ({}, {}, {})\n".format(system.name,
          system.coords['x'], system.coords['y'], system.coords['z'])
    return ret[:-1]

# ===========================================================================

parser = argparse.ArgumentParser(description="A collection of tools useful for "
    + "exploration.")
subparsers = parser.add_subparsers(title="subcommands", help="sub-command help",
    dest="subcommand", required=True)

parser_bodycount = subparsers.add_parser("bodycount",
    help="Returns the number of bodies in a system. Will exit with code 1 on "
    + "server error and code 2 if the system could not be found in EDSM.")
parser_bodycount.add_argument("system", nargs=1, help="system to query")

parser_distance = subparsers.add_parser("distancebetween",
    help="Calculates the distance between two systems. Will exit with code 1 "
    + "on server error and code 2 if (one of) the systems could not be found "
    + "on EDSM.")
parser_distance.add_argument("--roundto", nargs="?", type=int, default=2,
    help="the number of digits to round to (default: 2)")
parser_distance.add_argument("system", nargs=2, help="the systems to measure")

parser_findCmdr = subparsers.add_parser("findcommander",
    help="Attempts to find a CMDR’s last known position. Will exit with code 1 "
    + "on server error and code 2 if the CMDR could not be found on EDSM. Will "
    + "also give you the time of last activity if you search for their system.")
group = parser_findCmdr.add_mutually_exclusive_group(required=False)
group.add_argument('--system', action='store_true',
    help='output the commander’s last known system (default)')
group.add_argument('--coords', action='store_true',
    help='output the commander’s last known position in {x,y,z} coordinates')
group.add_argument('--url', action='store_true',
    help='output the commander’s profile URL')
parser_findCmdr.add_argument("name", help="the commander in question")
parser_findCmdr.add_argument("apikey", default="", nargs="?",
    help="the commander’s EDSM API key. Can be empty for public profiles.")

parser_findSystem = subparsers.add_parser("findsystem",
    help="Attempts to find a partially matching system that should then "
    + "hopefully be in the vicinity of the given system")
parser_findSystem.add_argument("system", help="the system in question")

parser_bodycount = subparsers.add_parser("systemlist",
    help="Pulls all system names starting with the given string from EDSM")
parser_bodycount.add_argument("partialsystem", nargs=1,
    help="the partial system name to query against")

argcomplete.autocomplete(parser)
args = parser.parse_args()

# ===========================================================================

try:
  if args.subcommand == "bodycount":
    out = getBodyCount(args.system[0])
  elif args.subcommand == "distancebetween":
    out = distanceBetween(args.system[0], args.system[1], args.roundto)
  elif args.subcommand == "findcommander":
    if args.coords:
      out = getCommanderPosition(args.name, args.apikey)
    elif args.url:
      out = getCommanderProfileUrl(args.name, args.apikey)
    else:
      out = getCommanderSystem(args.name, args.apikey)
  elif args.subcommand == "findsystem":
    out = getSystemNear(args.system)
  elif args.subcommand == "systemlist":
    out = getSystemList(args.partialsystem)
except ServerError as e:
  print(e)
  sys.exit(1)
except NotFoundError as e:
  print(e)
  sys.exit(2)
else:
  print(out)
  sys.exit(0)


spansh.py
#!/usr/bin/env python3
# PYTHON_ARGCOMPLETE_OK
import argcomplete, argparse
import json as JSON
import requests
import sys

from datetime import datetime, timedelta, timezone
from dateutil import parser as dtparser

from pyEDSM.edsm.exception import ServerError, NotFoundError

# ===========================================================================

def querystations(url, params):
  response = requests.post(url, params)
  if response.status_code != 200:
    raise ServerError(url, params)
  json = response.json()
  if json["count"] == 0:
    raise NotFoundError()
  return json

# ===========================================================================

def getNearestSystem(coords, raw=False):
  params = {
      "x": coords[0],
      "y": coords[1],
      "z": coords[2]
      }
  response = requests.post(APIURLS["nearest"], params)
  if response.status_code != 200:
    raise ServerError(url, params)

  json = response.json()

  if raw: return json

  ret = None
  system = json["system"]
  if args.short:
    ret = system["name"]
  elif args.parsable:
    ret = "{}|{},{},{}|{}".format(system["name"], system["x"],
        system["y"], system["z"], round(system["distance"], 2))
  else:
    system = json["system"]
    ret = "{} ({}, {}, {}), {} ly".format(system["name"], system["x"],
        system["y"], system["z"], round(system["distance"], 2))

  return ret

def getOldStations(raw=False):
  params = {
      "json": JSON.dumps({
        "filters": FILTERS,
        "sort": SORT,
        "size": args.count
        })
      }
  json = querystations(APIURLS["stations"], params)

  if raw: return json

  ret = ""
  for station in json["results"]:
    if args.short:
      ret += "{}\n".format(station["system_name"])
    else:
      ret += "{}: {} ({} days ago)\n".format(station["system_name"], station["name"],
        (datetime.now(timezone.utc) - dtparser.parse(station["updated_at"])).days)

  return ret[:-1]

def getOldStationsInSystem(system, raw=False):
  FILTERS.update(system_name={"value": system})
  params = {
      "json": JSON.dumps({
        "filters": FILTERS,
        "sort": SORT
        })
      }
  json = querystations(APIURLS["stations"], params)

  if raw: return json

  ret = ""
  if len(json["results"]) == 0:
    raise NotFoundError()
  for station in json["results"]:
    # systems including the given name as a word will also trigger;
    # looking for e.g. “Mari” will also give you stuff in “Mac Mari”!
    if station["system_name"] == system:
      if args.short:
        ret += "{}\n".format(station["name"])
      else:
        ret += "{} ({} days ago)\n".format(station["name"],
          (datetime.now(timezone.utc) - dtparser.parse(station["updated_at"])).days)

  return ret[:-1]

def systemExists(system, raw=False):
  params = {
      "json": JSON.dumps({
        "filters": {
          "name": {
            "value": system
            }
          }
        })
      }
  response = requests.post(APIURLS["systems"], params)
  if response.status_code != 200:
    raise ServerError(url, params)

  json = response.json()

  if raw: return json

  if json["count"] == 0:
    raise NotFoundError()

  ret = ""
  for system in json["results"]:
    ret += "{} ({}, {}, {})\n".format(system["name"], system["x"], system["y"],
        system["z"])

  return ret[:-1]

# ===========================================================================

parser = argparse.ArgumentParser(description="Script for interfacing with "
    + "Spansh’s API.")
parser.add_argument("--raw", action='store_true',
    help="output raw json (mostly for debugging)")
subparsers = parser.add_subparsers(title="subcommands", help="sub-command help",
    dest="subcommand", required=True)

parser_nearestsystem = subparsers.add_parser("nearestsystem",
    help="Searches for the nearest system in the database to given coordinates.")
parser_nearestsystem.add_argument("coordinate", nargs=3, type=int,
    help="the coordinates to search for (order: x, y, z)")
group = parser_nearestsystem.add_mutually_exclusive_group(required=False)
group.add_argument("--short", action='store_true',
    help="short output format (system name only)")
group.add_argument("--parsable", action='store_true',
    help="parsable output format (<name>|<x>,<y>,<z>|<distance>)")

parser_oldstations = subparsers.add_parser("oldstations",
    help="Searches for stations with old data (>1 year without an update.")
parser_oldstations.add_argument("--system", nargs="?",
    help="a single system to query. If not present, get the oldest stations "
    + "overall.")
parser_oldstations.add_argument("--count", nargs="?", type=int, default=50,
    help="how many stations to output. Defaults to 50.")
parser_oldstations.add_argument("--minage", nargs="?", type=int, default=365,
    help="minimum age of data (in days) to be considered “outdated”. Defaults to "
    + "365 (= 1 year).")
parser_oldstations.add_argument("--short", action='store_true',
    help="short output format (system/station names only)")
parser_oldstations.add_argument("--nofeet", action='store_true',
    help="omit Odyssey settlements")

parser_systemexists = subparsers.add_parser("systemexists",
    help="Checks if a given system exists in the search database.")
parser_systemexists.add_argument("system", nargs=1,
    help="the system to search for")

argcomplete.autocomplete(parser)
args = parser.parse_args()

# ===========================================================================

APIURLS = {
    "nearest": "https://spansh.uk/api/nearest",
    "stations": "https://spansh.uk/api/stations/search",
    "systems": "https://spansh.uk/api/systems/search",
    }

try:
  if args.subcommand == "nearestsystem":
    out = getNearestSystem(args.coordinate)
  elif args.subcommand == "oldstations":
    FILTERS = {
        "updated_at":
        {
          "value": [
            "2017-11-06",
            (datetime.now() - timedelta(days=args.minage)).strftime("%Y-%m-%d")
          ],
          "comparison": "<=>"
        },
        "type":
        {
          "value": [
            "Asteroid base",
            "Coriolis Starport",
            "Mega ship",
            "Ocellus Starport",
            "Orbis Starport",
            "Outpost",
            "Planetary Outpost",
            "Planetary Port",
            "Settlement" if not args.nofeet else ""
          ]
        }
      }
    SORT = {
        "updated_at": {
          "direction": "asc"
        }
      }
    if args.system:
      out = getOldStationsInSystem(args.system, args.raw)
    else:
      out = getOldStations(args.raw)
    if args.raw:
      out = JSON.dumps(out, indent=2)
  elif args.subcommand == "systemexists":
    out = systemExists(args.system, args.raw)
except ServerError as e:
  print(e)
  sys.exit(1)
except NotFoundError as e:
  print("No results found.")
  sys.exit(3)
else:
  print(out)
  sys.exit(0)
